x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.ts

```diff
@@ -0,0 +1,279 @@
+
+                def currentLocation = new HashMap(service);
+
+                /* only add the current location to the path if it's different from the last one*/
+                if (lastLocation == null || !lastLocation.equals(currentLocation)) {
+                  basePath.add(currentLocation);
+                }
+
+                /* if there is an outgoing span, create a new path */
+                if (event['destination.address'] != null && event['destination.address'] != '') {
+                  def outgoingLocation = getDestination(event);
+                  def outgoingPath = new ArrayList(basePath);
+                  outgoingPath.add(outgoingLocation);
+                  context.paths.add(outgoingPath);
+                }
+
+                event.path = basePath;
+
+                context.processedEvents[eventId] = event;
+                return event;
+              }
+
+              def context = new HashMap();
+
+              context.processedEvents = new HashMap();
+              context.eventsById = new HashMap();
+              
+              context.paths = new HashSet();
+              context.externalToServiceMap = new HashMap();
+              context.locationsToRemove = new HashSet();
+
+              for (state in states) {
+                context.eventsById.putAll(state);
+              }
+
+              for (entry in context.eventsById.entrySet()) {
+                processAndReturnEvent(context, entry.getKey());
+              }
+
+              def paths = new HashSet();
+
+              for(foundPath in context.paths) {
+                if (!context.locationsToRemove.contains(foundPath)) {  
+                  paths.add(foundPath); 
+                }
+              }
+
+              def response = new HashMap();
+              response.paths = paths;
+
+              def discoveredServices = new HashSet();
+
+              for(entry in context.externalToServiceMap.entrySet()) {
+                def map = new HashMap();
+                map.from = entry.getKey();
+                map.to = entry.getValue();
+                discoveredServices.add(map);
+              }
+              response.discoveredServices = discoveredServices;
+
+              return response;`
+            }
+          }
+        }
+      }
+    }
+  };
+
+  const serviceMapResponse = await client.search(serviceMapParams);
+
+  const scriptResponse = serviceMapResponse.aggregations?.service_map.value as {
+    paths: ConnectionNode[][];
+    discoveredServices: Array<{
+      from: ExternalConnectionNode;
+      to: ServiceConnectionNode;
+    }>;
+  };
+
+  let paths = scriptResponse.paths;
+
+  if (serviceName || environment) {
+    paths = paths.filter(path => {
+      return path.some(node => {
+        let matches = true;
+        if (serviceName) {
+          matches =
+            matches &&
+            'service.name' in node &&
+            node['service.name'] === serviceName;
+        }
```

Comment:
This actually doesn't take the discovered services (`destination.address => service.name`) into account. Not sure whether that's an issue though, and we'll never get it 100% right, because an earlier or later request might have discovered a service that this specific request doesn't see.